mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/StoreDocumentTournament.kt

```diff
@@ -134,11 +114,83 @@ class StoreDocumentTournament(
               .get<ChessDocument>()
               .toPoolResults()
-      createMatchesForRankedPlayers(results) { (round - 1).takeIf { it >= 1 } }
+      store.transaction { createMatchesForPoolResults(results) { (round - 1).takeIf { it >= 1 } } }
+    }
+  }
+
+  private fun Transaction<DocumentReference>.createPools(
+      tournamentDocumentReference: DocumentReference,
+      current: TournamentDocument,
+      maxPlayers: Int,
+      poolSize: Int,
+      bestOf: Int,
+      adminId: String,
+      players: List<String>
+  ) {
+    val chosenPlayers = players.shuffled().take(maxPlayers)
+    val pools =
+        chosenPlayers.chunked(poolSize).map { ids ->
+          ids.map { uid ->
+            uid to (get<ProfileDocument>(store.collection("users").document(uid))?.name ?: "")
+          }
+        }
+    val minOpponentsPerPool = pools.lastOrNull()?.size ?: 0
+
+    set(tournamentDocumentReference, current.copy(stage = StagePools))
+
+    pools
+        .asSequence()
+        .mapIndexed { index, participants ->
+          PoolDocument(
+              name = tournamentDetailsPoolName(index + 1),
+              tournamentId = reference.uid,
+              minOpponentsForAnyPool = minOpponentsPerPool,
+              remainingBestOfCount = bestOf,
+              tournamentBestOf = bestOf,
+              tournamentAdminId = adminId,
+              playerIds = participants.map { it.first },
+              playerNames = participants.map { it.second },
+          )
+        }
+        .forEach { set(store.collection(PoolDocument.Collection).document(), it) }
+  }
+
+  private fun Transaction<DocumentReference>.createFinalsMatchesForPlayers(
+      players: List<String>,
+      nextDepth: (TournamentDocument) -> Int?,
+  ) {
+    val ref = store.collection(Collection).document(reference.uid)
+    val currentDocument = get<TournamentDocument>(ref) ?: return
+
+    val bestOf = currentDocument.bestOf ?: return
+    val depth = nextDepth(currentDocument) ?: return
+    val count = 2.0.pow(depth).toInt()
+    val matches = players.take(count).chunked(2).filter { it.size == 2 }
+
+    set(
+        reference = ref,
+        value = currentDocument.copy(stage = stageDirectElimination(depth)),
+    )
+
+    for (match in matches) {
+      repeat(bestOf) { index ->
+        val (first, second) = if (index % 2 == 0) match[0] to match[1] else match[1] to match[0]
+        val matchRef = store.collection("games").document()
+        val matchDocument =
+            ChessDocument(
+                whiteId = first,
+                blackId = second,
+                lastUpdatedAt = System.currentTimeMillis(),
+                poolId = null,
+                roundDepth = depth,
+                tournamentId = document.uid,
+            )
+        set(matchRef, matchDocument)
+      }
     }
   }
 
-  private suspend fun createMatchesForRankedPlayers(
+  private fun Transaction<DocumentReference>.createMatchesForPoolResults(
```

Comment:
Missing some documentation for this method.


```kotlin
/**
 * Creates the matches corresponding to the given [PoolResults].
 *
 * @receiver the ongoing [Transaction].
 * @param results the [PoolResults] for which the matches are created.
 * @param nextDepth the function which computes the depth of the elimination tree.
 */
private fun Transaction<DocumentReference>.createMatchesForPoolResults()
```

Reply:
I did not know that we needed to document private functions?
Anyway, fixed in https://github.com/epfl-SDP/android/pull/392/commits/a8685cd7be6104b3a4f8315aabff2c43f026a76b